lib/AssemblyUtil/TypeToFasta.py

```diff
@@ -0,0 +1,80 @@
+
+
+import os
+
+from installed_clients.DataFileUtilClient import DataFileUtil
+from installed_clients.MetagenomeUtilsClient import MetagenomeUtils
+from AssemblyUtil.AssemblyToFasta import AssemblyToFasta
+
+from shutil import copyfile
+
+class TypeToFasta:
+
+    def __init__(self, callback_url, scratch, ws_url):
+        self.ws_url = ws_url
+        self.callback_url = callback_url
+        self.scratch = scratch
+        self.dfu = DataFileUtil(callback_url)
+        self.mgu = MetagenomeUtils(callback_url)
+
+    def type_to_fasta(self, ctx, ref_lst):
+
+        # Initiate objects
+        fasta_dict, fasta_array = dict(), []
+        atf = AssemblyToFasta(self.callback_url, self.scratch)
+
+        # Get type info for each ref in ref_lst
+        for idx, ref in enumerate(ref_lst):
+            # Initiate objects and get KBase object type with get_object_info3
+            upas = []
+            obj = {"ref": ref}
+            obj_info = self.ws_url.get_object_info3({"objects": [obj]})
+            obj_type = obj_info["infos"][0][2]
+
+            # Find type of object and get back KBase object with dfu.get_objects
+            if 'KBaseSets.GenomeSet' in obj_type:
+                obj_data = self.dfu.get_objects({"object_refs": [ref]})['data'][0]
+                upas = [gsi['ref'] for gsi in obj_data['data']['items']]
+            elif 'KBaseSearch.GenomeSet' in obj_type:
+                obj_data = self.dfu.get_objects({"object_refs": [ref]})['data'][0]
+                upas = [gse['ref'] for gse in obj_data['data']['elements'].values()]
+            elif "KBaseGenomes.Genome" in obj_type:
+                upas = [ref]
+            # If type Assembly, ContigSet or AssemblySet use assembly_as_fasta
+            elif "KBaseGenomes.ContigSet" in obj_type or "KBaseGenomeAnnotations.Assembly" in obj_type:
+                faf = [atf.assembly_as_fasta(ctx, obj)]
+                fasta_array.extend([faf[0]['path'], ref])
+
+            elif "KBaseSets.AssemblySet" in obj_type:
+                fasta_paths = []
+
+                obj_data = self.dfu.get_objects({"object_refs": [ref]})['data'][0]
+
+                for item_upa in obj_data['data']['items']:
+                    faf = [atf.assembly_as_fasta(ctx, {"ref": item_upa['ref']})]
+                    fasta_paths.extend([faf[0]['path'], item_upa['ref']])
+                    fasta_array = fasta_paths
+
+            elif 'KBaseMetagenomes.BinnedContigs' in obj_type:
+                fasta_paths = []
+
+                bin_file_dir = self.mgu.binned_contigs_to_file({'input_ref': ref, 'save_to_shock': 0})['bin_file_directory']
+                for (dirpath, dirnames, filenames) in os.walk(bin_file_dir):
+                    for fasta_file in filenames:
+                        fasta_path = os.path.join(self.scratch, fasta_file)
+                        copyfile(os.path.join(bin_file_dir, fasta_file), fasta_path)
```

Comment:
Doesn't MGU stick its files in scratch already?

Reply:
so the inner for loop is saying for every file in `bin_file_dir` which is computed from calling 'mgu.binned_contigs_to_file', there is a matching file in the staging area.
but the extension is  .fa. It then copies the .fa file in staging area and overwrites the original file.

Reply:
the inner for loop is just reseting the extension for every file in 'bin_file_dir' to '.fa'.

Reply:
It just copies all files matching the file name in the staging area. So if you have 'file.fa' in staging area and 'file.txt' in bin_file_dir, then it copies 'file.fa' from the staging area to bin_file_dir.

Reply:
What do you mean by `staging area`? That's a particular system in kbase that is used for staging files for upload/download which I don't think you're talking about here, but maybe I'm wrong.

I'm pretty lost at this point. I don't see where the `.fa` extension is being added in the code above - afaict `copyfile` is just copying a file from `bin_file_dir` into `self.scratch` - but `bin_file_dir`, unless `binned_contigs_to_file` is doing something very weird, should **already** be in `self.scratch`.

> So if you have 'file.fa' in the staging area and 'file.txt' in bin_file_dir, then it copies 'file.fa' from the staging area to bin_file_dir.

`copyfile` is `copyfile(src, dst)`...

Reply:
you're right I forgot to add the extension line..                          
fasta_path = os.path.splitext(fasta_path)[0] + ".fa"

for instance, if you have 'a.fa' in the staging area and 'a.txt' in 'bin_file_dir'. It copies 'a.fa' from staging area to 'bin_file_dir'
but if you have 'b.txt' in 'bin_file_dir' but no 'b.fa' in staging area. it will just keep the 'b.txt' in 'bin_file_dir'.

Reply:
I'm actually not confident the file is already in scratch and Tian is the one who instructed me to get the fasta_path this way for some reason...

Reply:
So Tian said "for every file in `bin_file_dir` which is computed from calling `mgu.binned_contigs_to_file`, there is a matching file in staging area.

but the extension is `.fa`."

Reply:
@tian can you elucidate here? Does `binned_contigs_to_file` really put stuff in the staging area? I don't think the function @CheyenneNS is writing will have access to it, necessarily, since it's meant for use in standard apps.

Reply:
it's the scratch not staging area staging area. Sorry cleared this up yesterday with @Tian.
I'm still trying to figure out the utility of this for loop.

Reply:
ok so no the fasta files aren't in scratch directly, their paths are /kb/module/work/tmp/binned_contig_files_be8662b9-34ce-44d3-bd6c-da918f09e648/Bin.001.fasta
and 
/kb/module/work/tmp/binned_contig_files_be8662b9-34ce-44d3-bd6c-da918f09e648/Bin.002.fasta

we then make a path directly to scratch:
 /kb/module/work/tmp/Bin.001.fasta and /kb/module/work/tmp/Bin.002.fasta

then we copy files from the bin_file_dir to /kb/module/work/tmp/

Reply:
all this for loop is doing is taking out fasta files from the binned_contig_files_be8662b9-34ce-44d3-bd6c-da918f09e648 folder and putting them in scratch. Changing the extension to .fa isn't needed.

Reply:
ok, so it's just moving the files up a level IIUC